Architecture-Proposal/Core/Fetcher.swift

```diff
@@ -0,0 +1,22 @@
+import RxSwift
+
+/**
+ If needed it will contains logical conditions used to constrain a fetch
+ */
+public protocol FetcherPredicate {}
+
+/**
+ Every Domain Model associated have a dedicated Fetcher
+ that is responsibile to retrieve the information.
+ The fetcher will always use a Service to fetch the model
+ */
+public protocol Fetcher {
+    associatedtype Model
+    func fetch(predicate: FetcherPredicate?) -> Observable<Model>
+}
+
+extension Fetcher {
+    func fetch(predicate: FetcherPredicate? = nil) -> Observable<Model> {
+        preconditionFailure("This method must be implemented in your fetcher or not used")
```

Comment:
this is a bit of a code smell, no? 😄 

I'm assuming that this extension is added so you get the default parameter. But then this also turns what _was_ a compile time error:

> fetch(predicate:) not implemented in class X declared in protocol Fetcher

into a runtime error (the precondition failure). Is the default parameter worth that?

A solution which doesn't have this issue is have a `fetch()` method in the extension, which calls through to`fetch(predicate: nil)`. That way, you still get the compile-time benefits _and_ a default version of the method without requiring a predicate:

```swift

extension Fetcher {
   func fetch() {
      fetch(predicate: nil)
   }
}

```

Reply:
yes, that could be a solution. 
do you think this method is providing any value to the Fetcher protocol? Can we be really always that generic in our concrete implementation of the Fetchers?

Reply:
no I don't think so - you should generally avoid generifying until you have at least 3 cases where it is useful. 

It depends how you see this - if it is a demo, make is the easiest it can be to get the point across.